pkg/reconciliation/decommission_node.go

```diff
@@ -54,7 +58,7 @@ func (rc *ReconciliationContext) DecommissionNodes(epData httphelper.CassMetadat
 		}
 	}
 
-	if currentSize <= dc.Spec.Size {
```

Comment:
I did a manual test with 2 DCs and 2 nodes per DC. When I deleted dc2, one node is decommissioned, but no more.  The pod is eventually recreated since the liveness check will fail and because the replica count on the statefulset isn't getting changed.

I haven't looked into it enough but it seems like it would be tricky to scale the StatefulSets down without also changing the size of the CassandraDatacenter. Definitely not a fan of that.

I am inclined to think that this might be easier to do from k8ssandra-operator. It would simply reduce the size of the dc, requeue until the decommission completes, reduce the size again, and so on until it hits zero. Other than allowing the last node to be decommissioned, would there be other changes required in cass-operator?

Reply:
After the decommission has succeeded, it will scale down StatefulSet. If your pod came up again it failed to decommission for some reason. Also, the CassandraDatacenter size is reduced during the operation.

Allowing user to set size to 0 would be bad, since it will never succeed. One can't call decommission to the last node if there are no other DCs. It will simply never finish.

Reply:
The scale down process itself was not changed, if it failed to reduce StS size then that issue is affecting all scale down operations

Reply:
The decommission did succeed. Then it looks like we requeue. The replica count of the StatefulSet doesn't change.

>Allowing user to set size to 0 would be bad, since it will never succeed. One can't call decommission to the last node if there are no other DCs. It will simply never finish.

But it's ok if we have another DC, right?

Reply:
If cass-operator sees decommission succeeding, it should remove the pod from StS:

https://github.com/k8ssandra/cass-operator/blob/master/pkg/reconciliation/decommission_node.go#L171

There's no requeue before doing this once it sees decommission has succeeded. Apparently it has missed the success for some reason. 

> But it's ok if we have another DC, right?

Assuming one rewrites the current rack scaling process, yes. The current rack scaling (up or down) is unable to function with size 0. It isn't simply a validation issue.

Reply:
Okay, I found the bug.